File: src/euler_problems/solution_2.clj
(ns euler-problems.solution-2
  (:use [util.misc]))

(defn digits
  ([n] (digits n 10))
  ([n ^long radix]
   (loop [n n
          res nil]
     (if (zero? n)
       res
       (recur
         (quot n radix)
         (cons (rem n radix) res))))))

(defn iexpt [x ^long exp]
  "Raise x to an integer exponent, of the form x^exp"
  (reduce * (repeat exp x)))

(defn sum-of-powers [pow n]
  (reduce + (map #(iexpt % pow) (digits n))))

(defn calc [coll]
  (reduce +
          (filter #(= % (sum-of-powers 5 %))
                  coll)))

(defn solve [batch-size limit]
  (reduce + (pmap calc (partition-all batch-size (range 32 limit)))))

(time (solve 15000 (* 6 (iexpt 9 5))))

File: src/euler_problems/Solution2.java
package euler_problems;

import java.util.stream.IntStream;

public class Solution2 {
    private static int fifthPowerDigitSum(int x) {
        int sum = 0;
        while (x != 0) {
            int y = x % 10;
            sum += y * y * y * y * y;
            x /= 10;
        }
        return sum;
    }

    public static void main(String[] args) {
        int sum = 0;
        for (int i = 2; i < 1000000; i++) {
            if (i == fifthPowerDigitSum(i))
                sum += i;
        }
        System.out.println(sum);

        int res = IntStream.range(2, 1000000)
                .filter(x -> x == fifthPowerDigitSum(x))
                .sum();

        IntStream.range(2, 1000000)
                .filter(x -> x == fifthPowerDigitSum(x))
                .sum();

        System.out.println(res);
    }

}
